compiler/src/main/java/com/squareup/anvil/compiler/InterfaceMerger.kt

```diff
@@ -87,143 +79,138 @@ internal class InterfaceMerger(
-          .annotationOrNull(contributesToFqName)
-        val contributesBindingAnnotation = classDescriptorForExclusion
-          .annotationOrNull(contributesBindingFqName)
-        val contributesMultibindingAnnotation = classDescriptorForExclusion
-          .annotationOrNull(contributesMultibindingFqName)
-        val contributesSubcomponentAnnotation = classDescriptorForExclusion
-          .annotationOrNull(contributesSubcomponentFqName)
+    val excludedClasses = mergeAnnotation.exclude()
+      .filter { it.isInterface() }
+      .onEach { excludedClass ->
+        val contributesToAnnotation = excludedClass
+          .annotations.find(contributesToFqName).singleOrNull()
+        val contributesBindingAnnotation = excludedClass
+          .annotations.find(contributesBindingFqName).singleOrNull()
+        val contributesMultibindingAnnotation = excludedClass
+          .annotations.find(contributesMultibindingFqName).singleOrNull()
+        val contributesSubcomponentAnnotation = excludedClass
+          .annotations.find(contributesSubcomponentFqName).singleOrNull()
 
         // Verify that the replaced classes use the same scope.
-        val scopeOfExclusion = contributesToAnnotation?.scope(module)
-          ?: contributesBindingAnnotation?.scope(module)
-          ?: contributesMultibindingAnnotation?.scope(module)
-          ?: contributesSubcomponentAnnotation?.parentScope(module)
-          ?: throw AnvilCompilationException(
-            thisDescriptor,
+        val scopeOfExclusion = contributesToAnnotation?.scope()
+          ?: contributesBindingAnnotation?.scope()
+          ?: contributesMultibindingAnnotation?.scope()
+          ?: contributesSubcomponentAnnotation?.parentScope()
+          ?: throw AnvilCompilationExceptionClassReference(
+            mergeAnnotatedClass,
             "Could not determine the scope of the excluded class " +
-              "${classDescriptorForExclusion.fqNameSafe}."
+              "${excludedClass.fqName}."
           )
 
-        if (scopeOfExclusion.fqNameSafe != scopeFqName) {
-          throw AnvilCompilationException(
-            thisDescriptor,
-            "${thisDescriptor.fqNameSafe} with scope $scopeFqName wants to exclude " +
-              "${classDescriptorForExclusion.fqNameSafe} with scope " +
-              "${scopeOfExclusion.fqNameSafe}. The exclusion must use the same scope."
+        if (scopeOfExclusion.fqName != mergeScope.fqName) {
+          throw AnvilCompilationExceptionClassReference(
+            mergeAnnotatedClass,
+            "${mergeAnnotatedClass.fqName} with scope ${mergeScope.fqName} wants to exclude " +
+              "${excludedClass.fqName} with scope " +
+              "${scopeOfExclusion.fqName}. The exclusion must use the same scope."
           )
         }
-
-        classDescriptorForExclusion.fqNameSafe
       }
-      ?: emptyList()
 
     if (excludedClasses.isNotEmpty()) {
       val intersect = supertypes.getAllSuperTypes()
```

Comment:
It would be nice if `getAllSuperTypes()` returns `ClassReference`. Maybe you can add a TODO?